File: python/xrand/tgfsr.py
class TGFSR:
    def __init__(self, seed=1):
        # Constants
        self.N = 25
        self.m = 7
        self.s = 7
        self.t = 15
        self.a = 0x8EBFD028
        self.b = 0x2B5B2500
        self.c = 0xDB8B0000

        # State variables
        self.x = [0] * self.N
        self.k = 0
        self.initialized = False
        
        # Initialize with seed
        self.initialize(seed)

    def initialize(self, seed):
        for i in range(self.N):
            seed &= 0xFFFFFFFF
            self.x[i] = seed
            seed = (seed * 1313 + 88897) & 0xFFFFFFFF
        self.k = self.N - 1
        self.initialized = True

    def iterate(self):
        for i in range(self.N - self.m):
            self.x[i] = self.x[i + self.m] ^ (self.x[i] >> 1) ^ (self.a if self.x[i] & 1 else 0)
        for i in range(self.N - self.m, self.N):
            self.x[i] = self.x[i + self.m - self.N] ^ (self.x[i] >> 1) ^ (self.a if self.x[i] & 1 else 0)

    def next(self):
        if not self.initialized:
            self.initialize(1)
        self.k += 1
        if self.k == self.N:
            self.iterate()
            self.k = 0
        y = self.x[self.k] ^ ((self.x[self.k] << self.s) & self.b)
        y ^= (y << self.t) & self.c
        return y & 0xFFFFFFFF

    def random(self):
        return self.next() & 0x7FFFFFFF

File: go/xrand/tgfsr/tgfsr.go
package tgfsr

type TGFSR struct {
    x           [25]uint32
    k           int
    initialized bool

    // Constants
    n int
    m int
    s int
    t int
    a uint32
    b uint32
    c uint32
}

func New() *TGFSR {
    return &TGFSR{
        n: 25,
        m: 7,
        s: 7,
        t: 15,
        a: 0x8ebfd028,
        b: 0x2b5b2500,
        c: 0xdb8b0000,
    }
}

func (tg *TGFSR) Initialize(seed uint32) {
    for i := 0; i < tg.n; i++ {
        seed &= 0xffffffff
        tg.x[i] = seed
        seed = (seed*1313 + 88897) & 0xffffffff
    }
    tg.k = tg.n - 1
    tg.initialized = true
}

func (tg *TGFSR) iterate() {
    for i := 0; i < tg.n-tg.m; i++ {
        tg.x[i] = tg.x[i+tg.m] ^ (tg.x[i] >> 1) ^ (func() uint32 {
            if tg.x[i]&1 != 0 {
                return tg.a
            }
            return 0
        }())
    }
    for i := tg.n - tg.m; i < tg.n; i++ {
        tg.x[i] = tg.x[i+tg.m-tg.n] ^ (tg.x[i] >> 1) ^ (func() uint32 {
            if tg.x[i]&1 != 0 {
                return tg.a
            }
            return 0
        }())
    }
}

func (tg *TGFSR) Next() uint32 {
    if !tg.initialized {
        tg.Initialize(1)
    }
    tg.k++
    if tg.k == tg.n {
        tg.iterate()
        tg.k = 0
    }
    y := tg.x[tg.k] ^ ((tg.x[tg.k] << tg.s) & tg.b)
    y ^= (y << tg.t) & tg.c
    return y
}

func (tg *TGFSR) Random() int32 {
    return int32(tg.Next() & 0x7fffffff)
}